javascripts/discourse-kanban/components/discourse-kanban.js.es6

```diff
@@ -9,21 +9,132 @@ export default Ember.Component.extend({
+            .split("|")
+            .map(m => m.split(":"));
+        const lookup = category || tag || "@";
+        const defaultMode = setDefaults.find(m => m[0] === lookup);
+
+        // triggered when in tag discovery mode, with tag columns/lists
+        if (
+            tag &&
+            definitions === "async_lookup_tags"
+        ) {
+            const asyncDefinitions_unfiltered = [];
+            this.store.findAll("tag").then(results => {
+
+                results.content.forEach(t => {
+                    asyncDefinitions_unfiltered.push({
+                        title: `#${t.id}`,
+                        params: {
+                            tags: t.id
+                        }
+                    });
+                });
+                this.filterTagDefinitions(asyncDefinitions_unfiltered);
+            });
+        }
+
+        // only trigger column filtering when there are no pre-defined columns to display (via settings) and when we are in tag discovery mode, displaying category columns
+        else if (
+            definitions &&
+            tag && // user building a Kanban board in tag discovery mode
+            ((definitions[0].params.category && // user building lists out of categories (i.e. category key exists)
+            !defaultMode // user not set a default mode list of categories to display for this board
+            ) || board === "categories") // OR user has changed settings modal to display categories, without specifying which ones
+        ) {
+            console.log('test')
+            this.filterTagDefinitions(definitions);
+        }
+
+        // base case
+        else {
+            console.log('fail')
+            this.set("definitions", definitions);
+        }
+    },
+
+    /**
+     * Helper function for tag discovery mode to query db for topics under each list
+     * then removes empty lists from definitions array:
+     *  1) category lists (d.params.category is truthy):
+     *     queries topics under each category list for the current discovery tag
+     *  2) tag lists (d.params.tags is truthy):
+     *     queries topics under each tag list... then filters offline to only keep topics that include current
+     *     discovery tag (unable to find method for querying topics filtered by more than 1 tag).
+     */
+    filterTagDefinitions(definitions_unfiltered) {
+        const tag = this.kanbanHelper.discoveryTag;
+        const promises = definitions_unfiltered.map(d => {
+
+            const params = {};
+            if (d.params.category) params.category = d.params.category;
+            if (d.params.tags) params.tag = d.params.tags;
+            
+            const storeOptions = {};
+            storeOptions.params = params;
+            if (d.params.category) storeOptions.filter = `tags/${tag.id}`
+            if (d.params.tags) storeOptions.filter = `tags/${d.params.tags}`
+
+            return this.store.findFiltered("topicList", storeOptions);
```

Comment:
This is going to cause a very large number of AJAX requests to the server, and probably trigger rate limiting.  I suspect this could be the cause of your issue [here](https://publichappinessmovement.com/t/topic/1764). I don't think the automatic removal of empty lists is feasible without causing rate limiting issues. I think it would be better to let users specify specific categories, or just display all categories.